pkmn_env.py

```diff
@@ -1,91 +1,43 @@
-import asyncio
 import time
 import threading
 import os
+import subprocess
 
 class Env():
     def __init__(self):
-        self.proc = ""
         self.done = True
-        self.lock = ""
-        self.observationBuffer = ""
     def seed(self, num):
         dummy = True
     def render(self):
         print(notDefinedYet)
     def reset(self):
-        # Create a Pokemon battle in a separate thread.
-        self.lock = threading.Lock()
-        self.loop = asyncio.new_event_loop()
-        asyncio.get_child_watcher().attach_loop(self.loop)
-        self.t = threading.Thread(target = self.pokemon_wrapper, args = (self.loop,))
+        # Create a Pokemon battle.
+        self.t = threading.Thread(target = self.pokemon_wrapper)
         self.t.start()
         # Wait a bit for the game to initialise.
         time.sleep(2)
         # The Pokemon battle will return some stuff
         return self.scrape_input()
     def step(self, action):
-        # I'm trying to write to standard input in various ways here. This is because if
-        # you type in to standard input from the keyboard, it works.
-
-        # Attempt 1, open file descriptor 0 (which is standard input).
-        # Gives no errors, also doesn't work.
-        #stin=open(0,"w")
-        #stin.write(action + "\n")
-        #stin.close()
-        
-        # Attempt 2, tell the operating system to print the action.
-        # This produces no errors, but it also doesn't work.
-        #os.system("echo " + action)
-
-        # Attempt 3, communicate with the process directly.
-        # Communicate, like all of process's methods, is asynchronous, so this gives an error.
-        #self.proc.communicate(action.encode())
-        # This gives a "loop is being detached from a child watcher with pending handlers" warning
-        # and an "attached to a different loop" error. (this attachment is probably in run).
-        #asyncio.run(self.proc.communicate(action.encode()))
-
-        # Attempt 4, communicate with the process directly in a different thread.
-        # This also gives an attached to a different loop error.
-        #myloop = asyncio.new_event_loop()
-        #mythread = threading.Thread(target = self.wr, args = (myloop, action,))
-        #mythread.start()
-
+        # Obscure method of writing to user input.
+        for letter in action:
+          if letter == ' ':
+            os.system('xdotool key space')
+          else:
+            os.system('xdotool key ' + letter)
+        os.system('xdotool key Return')
         # Wait for the Pokemon simulator to do its thing. This is not very efficient
-        # but it does allow human input (which works).
-        time.sleep(3)
+        # but it does allow human input.
+        time.sleep(2)
```

Comment:
This sleep will probably make it difficult to do RL, since each step is going to take that long. How did https://github.com/yuzeh/metagrok do the interfacing?

Reply:
How much do you want to know the answer? I cloned the repository and took a look at it. It might take a day to answer the question? Mostly in the form of understanding big pieces of software I haven't seen before (docker, pytorch). Doable, not nothing.
    (For reference, I would ballpark that the game engine takes 0.01 seconds to process a turn. (I recall it took about 0.65 seconds a battle for random battles, and 65 turns per battle seems about right). In the near future, that time.sleep is going to get cut by two orders of magnitude, and in the further future perhaps more).

Reply:
http://cs230.stanford.edu/projects_fall_2018/reports/12447633.pdf says "HTTP requests and responses are used to bridge the gap between our primary language (Python) and the Pokemon Showdown simulator environment (Javascript)". Yep, yep that's certainly one way to do it.